### PR TITLE
enforce order of user-defined tasks at launch

### PR DESCRIPTION
**Pull request by davidcoutadeur**
See #67, and especially https://github.com/lsc-project/lsc/issues/67#issuecomment-1597310928

### CODE REVIEW

**Review by rouazana — COMMENTED** [review]
Hello,

Thanks for trying to fix issue! I've just experienced it by the way.

The easiest workaround is to match the order of your task with the alphabetic order:
`1-first-task, 2-second-task, 3-third-task, etc.`

But I agree we should change the code for this so that in the next version tasks are executed in the same order they are submitted by the user.

I guess your fix should work but it seems a little over complicated. I think a more simpler solution would be [at this line](https://github.com/lsc-project/lsc/commit/515e459549b8f96b3dac47a42a7af8ccc2a286bc#diff-d5b6d928746bbfe34818c3bd0b85a5a23137c0d41e3a8ee98296422091e9bcf8R97) to replace TreeMap by a LinkedHashMap. It should do the job in a much simpler way.

Could you check it's working as expected?

I think it would also be worth to add a test to check and enforce this behavior in the future.

Thank you,
Raphaël.

**Comment by davidcoutadeur** [comment]
Hi!
Thanks for your help. :) I'll try your suggestion quite soon.

**Comment by davidcoutadeur** [comment]
For info, I have also tested your suggestion:

```
diff --git a/src/main/java/org/lsc/SimpleSynchronize.java b/src/main/java/org/lsc/SimpleSynchronize.java
index aa8f6d63..6943471d 100644
--- a/src/main/java/org/lsc/SimpleSynchronize.java
+++ b/src/main/java/org/lsc/SimpleSynchronize.java
@@ -57,6 +57,7 @@ import java.util.Map;
 import java.util.Map.Entry;
 import java.util.Set;
 import java.util.TreeMap;
+import java.util.LinkedHashMap;
 
 import org.apache.commons.lang.ArrayUtils;
 import org.lsc.beans.IBean;
@@ -95,7 +96,7 @@ public class SimpleSynchronize extends AbstractSynchronize {
        public SimpleSynchronize() {
                super();
                setThreads(5); 
-               cache = new TreeMap<String, Task>();
+               cache = new LinkedHashMap<String, Task>();
        }
        
        public void init() throws LscConfigurationException {

```

For me, it does not work, because when I launch:

```
lsc -f ~/Documents/lsc/etc -sMySyncTask3,MySyncTask2,MySyncTask1 -cMySyncTask3,MySyncTask2,MySyncTask1 -t 1
```

The order is not enforced:

```
juin 29 11:03:44 - INFO  - Starting sync for MySyncTask1
juin 29 11:03:45 - INFO  - All entries: 8, to modify entries: 0, successfully modified entries: 0, errors: 0
juin 29 11:03:45 - INFO  - Starting clean for MySyncTask1
juin 29 11:03:45 - INFO  - All entries: 8, to modify entries: 0, successfully modified entries: 0, errors: 0
juin 29 11:03:45 - INFO  - Starting sync for MySyncTask2
juin 29 11:03:45 - INFO  - All entries: 8, to modify entries: 0, successfully modified entries: 0, errors: 0
juin 29 11:03:45 - INFO  - Starting clean for MySyncTask2
juin 29 11:03:45 - INFO  - All entries: 8, to modify entries: 0, successfully modified entries: 0, errors: 0
juin 29 11:03:45 - INFO  - Starting sync for MySyncTask3
juin 29 11:03:46 - INFO  - All entries: 8, to modify entries: 0, successfully modified entries: 0, errors: 0
juin 29 11:03:46 - INFO  - Starting clean for MySyncTask3
juin 29 11:03:46 - INFO  - All entries: 8, to modify entries: 0, successfully modified entries: 0, errors: 0
```

But maybe I didn't understand your point?

**Comment by rouazana** [comment]
Your are probably right. I guess my proposed fix takes the tasks in the order they are in the configuration file, but still not as specified on the command line.

**Comment by rouazana** [comment]
I you can add a test it would be great, else I'm fine with your fix.

Also I would have argued that it would have been better to extract methods instead of copying code, but it seems it will be hard given all the boolean that are manipulated.

**Comment by davidcoutadeur** [comment]
> Also I would have argued that it would have been better to extract methods instead of copying code, but it seems it will be hard given all the boolean that are manipulated.

I really tried believe me :)

I'll take care of the test

**Comment by rouazana** [comment]
Just to notice: your fix introduces a slightly different behavior.
For example with `-s t1,t2 -c t1,t2` before LSC was doing:
```
sync t1, clean t1, sync t2, clean t2
```
If I got it right with your fix it will do:
```
sync t1, sync t2, clean t1, clean t2
```

I'm not sure if an issue, but in case you did not notice it...

**Comment by rouazana** [comment]
Also here should be a slightly simpler fix:
```
diff --git a/src/main/java/org/lsc/SimpleSynchronize.java b/src/main/java/org/lsc/SimpleSynchronize.java
index 4e5a0835..61184599 100644
--- a/src/main/java/org/lsc/SimpleSynchronize.java
+++ b/src/main/java/org/lsc/SimpleSynchronize.java
@@ -57,6 +57,7 @@ import java.util.Map;
 import java.util.Map.Entry;
 import java.util.Set;
 import java.util.TreeMap;
+import java.util.stream.Collectors;
 
 import org.apache.commons.lang3.ArrayUtils;
 import org.lsc.beans.IBean;
@@ -138,16 +139,10 @@ public class SimpleSynchronize extends AbstractSynchronize {
 	 */
 	public final boolean launch(final List<String> asyncTasks, final List<String> syncTasks,
 					final List<String> cleanTasks) throws Exception {
-		Boolean foundATask = false;
+		boolean foundATask = false;
 		boolean canClose = true;
 		boolean launchResult = true;
 
-		// Get the list of defined tasks from LSC properties
-		// Iterate on each task
-		boolean isASyncTaskAll = asyncTasks.contains(ALL_TASKS_KEYWORD);
-		boolean isSyncTaskAll = syncTasks.contains(ALL_TASKS_KEYWORD);
-		boolean isCleanTaskAll = cleanTasks.contains(ALL_TASKS_KEYWORD);
-		
 		if(getTasksName() == null) {
 			return false;
 		} else if(getTasks().length == 0) {
@@ -156,43 +151,43 @@ public class SimpleSynchronize extends AbstractSynchronize {
 		if(!asyncTasks.isEmpty()) {
             LscServerImpl.startJmx(this);
 		}
-		
-		for (Task task: cache.values()) {
 
-			// Launch the task either if explicitly specified or if "all" magic keyword used
-			if (isSyncTaskAll || syncTasks.contains(task.getName())) {
-				foundATask = true;
+		Collection<Task> asyncTasksToLaunch = userOrderedTask(asyncTasks, asyncTasks.contains(ALL_TASKS_KEYWORD));
+		Collection<Task> syncTasksToLaunch = userOrderedTask(syncTasks, syncTasks.contains(ALL_TASKS_KEYWORD));
+		Collection<Task> cleanTasksToLaunch = userOrderedTask(cleanTasks, cleanTasks.contains(ALL_TASKS_KEYWORD));
 
-				if (!launchTask(task, Task.Mode.sync)) {
-					launchResult = false;
-				} else {
-					if(task.getSyncHook() != null && task.getSyncHook() != "") {
-						runPostHook(task.getName(), task.getSyncHook(), task.getTaskType());
-					}
+		for (Task task: syncTasksToLaunch) {
+			foundATask = true;
+
+			if (!launchTask(task, Task.Mode.sync)) {
+				launchResult = false;
+			} else {
+				if(task.getSyncHook() != null && task.getSyncHook() != "") {
+					runPostHook(task.getName(), task.getSyncHook(), task.getTaskType());
 				}
 			}
-			if (isCleanTaskAll || cleanTasks.contains(task.getName())) {
-				foundATask = true;
+		}
+		for (Task task: cleanTasksToLaunch) {
+			foundATask = true;
 
-				if (!launchTask(task, Task.Mode.clean)) {
-					launchResult = false;
-				} else {
-					if(task.getCleanHook() != null && task.getCleanHook() != "") {
-						runPostHook(task.getName(), task.getCleanHook(), task.getTaskType());
-					}
+			if (!launchTask(task, Task.Mode.clean)) {
+				launchResult = false;
+			} else {
+				if(task.getCleanHook() != null && task.getCleanHook() != "") {
+					runPostHook(task.getName(), task.getCleanHook(), task.getTaskType());
 				}
 			}
-			if (isASyncTaskAll || asyncTasks.contains(task.getName())) {
-				foundATask = true;
-				
-				canClose = false;
+		}
+		for (Task task: asyncTasksToLaunch) {
+			foundATask = true;
 
-				if(!launchTask(task, Task.Mode.async)) {
-					launchResult = false;
-				}
+			canClose = false;
+
+			if(!launchTask(task, Task.Mode.async)) {
+				launchResult = false;
 			}
 		}
-		
+
 		if (canClose) {
 			close();
 		}
@@ -205,6 +200,18 @@ public class SimpleSynchronize extends AbstractSynchronize {
 		return launchResult;
 	}
 
+	private Collection<Task> userOrderedTask(List<String> userValues, boolean all) {
+		Collection<Task> allTasks = cache.values();
+		if (all) {
+			return allTasks;
+		}
+		Collection<String> allTaskNames = allTasks.stream().map(Task::getName).collect(Collectors.toList());
+		return userValues.stream()
+				.filter(allTaskNames::contains)
+				.map(cache::get)
+				.collect(Collectors.toList());
+	}
+
 	/**
 	 * Launch a task. Call this for once each task type and task mode.
 	 *
```
(exactly the same behavior than yours)

**Comment by davidcoutadeur** [comment]
> Just to notice: your fix introduces a slightly different behavior. For example with `-s t1,t2 -c t1,t2` before LSC was doing:
> 
> ```
> sync t1, clean t1, sync t2, clean t2
> ```
> 
> If I got it right with your fix it will do:
> 
> ```
> sync t1, sync t2, clean t1, clean t2
> ```
> 
> I'm not sure if an issue, but in case you did not notice it...

Yes, I already had that in mind. It does not seem a big change, and seems quite logical.

Anyway, as this is a modification of behaviour, there is a discussion about "in which version we could apply this fix". I mean that really enforcing the task order *is* a real modification of the behaviour, even if the doc says the contrary.

Thus it could be applied to the future major release maybe?

**Comment by davidcoutadeur** [comment]
> ...
> (exactly the same behavior than yours)

Seems more elegant indeed. I'll try this and add the test this week.

**Comment by davidcoutadeur** [comment]
@rouazana I have applied your suggestion + written the test for enforcing the task order.

Could you review the test? If it is ok for you, I think we can merge.

File: src/test/java/org/lsc/Ldap2LdapOrderedSyncTest.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+		// check the results of the synchronization
+		reloadJndiConnections();
+		checkSyncResults();
+	}
+
+	private final void checkSyncResults() throws Exception {
+		List<String> attributeValues = null;
+
+
+		// check Operation order
```
Comment (rouazana) [inline src/test/java/org/lsc/Ldap2LdapOrderedSyncTest.java:116]: it seems your test is not checking anything ^^

File: src/test/java/org/lsc/Ldap2LdapOrderedSyncTest.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+import org.lsc.jndi.SimpleJndiSrcService;
+import org.lsc.service.IService;
+import org.lsc.utils.directory.LDAP;
+
+/**
+ * This test case attempts to reproduce a typical ldap2ldap setup via
+ * SimpleSynchronize. This relies on settings in lsc.properties from the test
+ * resources, and entries in the local OpenDS directory. testSync() performs a
+ * synchronization between two branches of the local LDAP server, and should
+ * perform 3 operations : 1 ADD, 1 MODRDN, 1 MODIFY. The
```
Comment (rouazana) [inline src/test/java/org/lsc/Ldap2LdapOrderedSyncTest.java:82]: copy/pasted comment?

**Review by rouazana — COMMENTED** [review]
it seems your test is not ensuring the order is enforced.

File: src/test/java/org/lsc/Ldap2LdapOrderedSyncTest.java (hunk 1 above)
Comment (davidcoutadeur) [reply src/test/java/org/lsc/Ldap2LdapOrderedSyncTest.java:116]: The test itself just launch the tasks in a specific order (3,2,1), while these tasks are defined in an other order in lsc.xml (1,2,3).
But when the tasks are executed in the order (1,2,3), it ends inevitably in an error. (see lsc.xml content)
The only way the tasks can succeed is to execute in order (3,2,1)

File: src/test/java/org/lsc/Ldap2LdapOrderedSyncTest.java (hunk 2 above)
Comment (davidcoutadeur) [reply src/test/java/org/lsc/Ldap2LdapOrderedSyncTest.java:82]: Yes, duplicated from `Ldap2LdapSyncTest.java`

File: src/test/java/org/lsc/Ldap2LdapOrderedSyncTest.java (hunk 1 above)
Comment (davidcoutadeur) [reply src/test/java/org/lsc/Ldap2LdapOrderedSyncTest.java:116]: The other idea I had was to capture the lsc output to check the real order execution, but it seemed more complex to me.

Comment (rouazana) [reply src/test/java/org/lsc/Ldap2LdapOrderedSyncTest.java:116]: ohhhh I totally missed this.

I think it's ok if you add two comments:
1. replace the [main comment](https://github.com/lsc-project/lsc/pull/218#discussion_r1254837017) by an explanation
2. replace `// check Operation order` by a little explanation telling why it fails

Comment (davidcoutadeur) [reply src/test/java/org/lsc/Ldap2LdapOrderedSyncTest.java:113]: I have fixed your 2 last points. I will merge the PR

**Comment by rouazana** [comment]
Great thank you!

**Pull request merged**